rfd/0194-selinux-policy.md

```diff
@@ -0,0 +1,150 @@
+
+## Why
+
+Teleport is often deployed in environments with strict security requirements. SELinux provides a way to enforce fine-grained access controls, which
+can help prevent privilege escalation and lateral movement, provide peace of mind, and comply with security standards. As agents are the most widely
+deployed component of Teleport, it is important to ensure that they are secure and compliant.
+
+Only SSH agents are initially in scope for this proposal as a starting point, but the SELinux policy may be extended to other components of Teleport in the future.
+
+## Details
+
+### SELinux Policy Structure
+
+SSH agents need a lot of privileges to function properly. They need to bind to privileged ports, manage cgroups, create processes, and more.
+When an SSH session is started, the SSH service will create 2 child processes to handle it: 
+
+- An intermediary process that prepares another child process that will actually run the SSH session. Depending on what is configured, PAM contexts
+may be created, users may be created, a shell will be created, the user and group will be changed if necessary, and so on.
+- The intermediary process will then create the child session process, which only needs standard permissions of the host user that the SSH session
+was created for.
+
+Additionally, the SSH agent may create a networking child process to handle X11 forwarding, port forwarding, and SSH Agent forwarding. This process
+is a child process of the intermediary process as well.
+
+4 main SELinux domains will be created to ensure each process in the chain is granted only the necessary permissions: `teleport_ssh_service`,
+`teleport_ssh_service_intermediary`, `teleport_ssh_service_session`, and `teleport_ssh_service_networking` respectively.
+
+Configurable file and port contexts will be created for the `teleport_ssh_service` domain, and configurable file contexts
+will be created for the `teleport_ssh_service_intermediary` domain to allow those processes to bind to the necessary ports and read and write
+to the necessary files, depending on the SSH agent configuration. At a minimum this will allow the `teleport_ssh_service` domain to bind to
+the configured SSH port and read and write to the data directory, and dial the configured Teleport Proxy and Auth servers. 
+
+There are a few optional features that can be enabled or disabled on SSH agents, the first version of the policy will ensure necessary permissions
+for all optional features are allowed to simplify development and deployment.
+
+These include:
+
+- Reading `~/.tsh/environment` on the server before creating a session
+    - The `teleport_ssh_service_intermediary` domain will be allowed to read `.tsh/environment` in all users' home directories
+- Creation of host users and groups
+    - The `teleport_ssh_service` domain will be allowed to call necessary binaries such as `useradd`, `usermod` and `groupadd`
```

Comment:
Not sure if there is a way around it (at least with selinux alone), but this provides an escape hatch if the process is assumed to be compromised. If `useradd` or `usermod` can be called by the process, then the process can reconfigure the user's shell, causing a different process to run when a user logs in (via SSH or any other means).

Reply:
Yeah I don't easily know of a simple way to enforce this, as SELinux doesn't allow filtering based off of arguments to a process. I can basically only control what permissions to what files (almost everything is a file in Linux) are allowed if you boil it down. That's a good thought though. There are a couple of other permissions, capabilities mainly that I also have to allow for Teleport to function correctly that could also be abused as well, no real way around that.